bpo-34850: Emit a warning for "is" and "is not" with a literal.

serhiy-storchaka opened the pull request:
<!-- issue-number: [bpo-34850](https://www.bugs.python.org/issue34850) -->
https://bugs.python.org/issue34850
<!-- /issue-number -->

serhiy-storchaka committed:
bpo-34850: Emit a warning for "is" and "is not" with a literal.

serhiy-storchaka committed:
Remove unintentionally inserted character.

gpshead commented on Doc/whatsnew/3.8.rst:
```diff
@@ -305,6 +305,12 @@ Changes in Python behavior
   (Contributed by Serhiy Storchaka in :issue:`10544`.)
 
+* The compiler will emit a syntax warning when the ``is`` or ``is not``
+  operator is used with a literal (except singletons ``None``, ``False``,
```
It is normally incorrect for code to perform an identity check against `True` or `False`.

serhiy-storchaka replied:
It is normal in the code that serializes Python data to external format that supports booleans. For example JSON, plist or pickle.

Since there are valid use cases for this, the compiler shouldn't emit a warning. Third-party checkers can do this if this check is optional.

warsaw replied:
Oh, I've definitely seen it more than a few times. :(

warsaw replied:
Identity checks against `True` and `False` are definitely code smells.  I always call them out in reviews.  But they are also fairly rare and as @serhiy-storchaka points out, do have some valid use cases.

gpshead commented on Python/compile.c:
```diff
@@ -2155,6 +2156,41 @@ compiler_class(struct compiler *c, stmt_ty s)
+    }
+    PyObject *value = e->v.Constant.value;
+    return (value == Py_None
+         || value == Py_False
```
I wouldn't special case True and False, those are not normally identity checked.  I don't doubt that _someone_ may have create APIs returning multiple possible types that requires this to differentiate between "real" values and bools... but yuck.

zhangyangyu commented on Lib/test/test_grammar.py:
```diff
@@ -1099,6 +1100,14 @@ def testAssert2(self):
+            warnings.filterwarnings('error', category=SyntaxWarning)
+            with self.assertRaisesRegex(SyntaxError, 'assertion is always true'):
+                compile('assert(x, "msg")', '<testcase>', 'exec')
+            compile('assert x, "msg"', '<testcase>', 'exec')
```
Whoo, `compile('assert(x, "msg")')` raise `SyntaxWarning` while `compile('assert(1, "msg")')` not. Interesting ast optimization.

serhiy-storchaka replied:
It is possible to add a support for a constant tuple, but I think this is not worth.

warsaw replied:
Agreed.

warsaw commented on Python/compile.c:
```diff
@@ -2155,6 +2156,41 @@ compiler_class(struct compiler *c, stmt_ty s)
+            if (!right || !left) {
+                const char *msg = (op == Is)
+                        ? "\"is\" with a literal. Did you mean \"==\""
+                        : "\"is not\" with a literal. Did you mean \"!=\"";
```
I think this warning looks good.  I'd like to, but can't come up with a good way to work in the phrase "identity check".

serhiy-storchaka replied:
Do you have suggestions for the wording of the news or What's New entry?

warsaw replied:
How about: 

>The compiler now produces a SyntaxWarning when identity checks (`is` and `is not`) are used with string literals.  These can often work by accident in CPython, but are not guaranteed by the language spec.  The warning advises users to use equality tests (`==` and `!=`) instead.

serhiy-storchaka replied:
Thanks! But a warning will be emitted not just for string literals, but for numerical literals and tuples (empty tuple is a singleton in CPython, this is an implementation detail).

warsaw replied:
You can say "certain types of literals (e.g. strings, ints)" rather than "string literals".

terryjreedy committed:
Update iomenu.py

Revert change here; will make separate PR to be applied and backported regardless of fate of this PR.

serhiy-storchaka committed:
Merge branch 'master' into is-with-literal-warning

serhiy-storchaka committed:
Update documentation and add missed "?".

serhiy-storchaka committed:
Merge branch 'master' into is-with-literal-warning

serhiy-storchaka committed:
Add comments.
